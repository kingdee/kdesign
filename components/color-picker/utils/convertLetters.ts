export function toLowerCase(str: string): string {
  let result = ''
  for (let i = 0; i < str.length; i++) {
    const charCode = str.charCodeAt(i)
    if (charCode >= 65 && charCode <= 90) {
      result += String.fromCharCode(charCode + 32)
    } else {
      result += str[i]
    }
  }
  return result
}
