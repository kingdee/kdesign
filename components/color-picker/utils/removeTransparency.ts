import { ColorTypes } from '../interface'
import { colorFormat } from './colorFormat'

export type ICurrentColorType = Exclude<typeof ColorTypes[number], 'themeColor'>

export const removeTransparency = (color: string, currentColorType: ICurrentColorType) => {
  return colorFormat(color, 1, currentColorType, true) as string
}
