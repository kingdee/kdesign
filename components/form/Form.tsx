import React, { useContext, useEffect } from 'react'
import classnames from 'classnames'
import { ConfigContext } from '../config-provider'
import { getCompProps } from '../_utils'
import { Callbacks, FormLayout, FormInstance, LabelAlign, Store } from './interface'
import useForm, { INTERNAL_HOOK_KEY } from './hooks/useForm'
import FormContext from './FormContext'
import FieldContext from './FieldContext'

type BaseFormProps = Omit<React.FormHTMLAttributes<HTMLFormElement>, 'onSubmit'>

export interface FormProps<Values = any> extends BaseFormProps {
  children?: React.ReactNode
  defaultValues?: Store // 默认值
  disabled?: boolean // 是否禁用
  className?: string // 类名
  form?: FormInstance<Values> // 表格引用
  labelAlign?: LabelAlign // 标签对齐方式
  labelWidth?: string | number // 标签宽度
  layout?: FormLayout // 布局
  name?: string // 表单名称
  wrapperWidth?: string | number // 输入区域宽度
  onFinish?: Callbacks<Values>['onFinish'] // 表单提交且数据验证通过
  onFinishFailed?: Callbacks<Values>['onFinishFailed'] // 表单提交且数据验证失败
  onValuesChange?: Callbacks<Values>['onValuesChange'] // 字段值更新
}

const Form: React.ForwardRefRenderFunction<FormInstance, FormProps> = (props: FormProps, ref) => {
  const {
    getPrefixCls,
    prefixCls,
    compDefaultProps: userDefaultProps,
    locale: globalLocale,
    direction,
  } = useContext(ConfigContext)
  const formProps = getCompProps('Form', userDefaultProps, props)

  const formLang = globalLocale.getCompLangMsg({ componentName: 'Form' })

  const {
    children,
    defaultValues,
    className,
    form,
    labelAlign,
    labelWidth,
    layout,
    name,
    prefixCls: customizePrefixCls,
    wrapperWidth,
    onFinish,
    onFinishFailed,
    onValuesChange,
    disabled,
    ...restProps
  } = formProps

  const formPrefixCls = getPrefixCls!(prefixCls, 'form', customizePrefixCls)
  const rtlCls = direction === 'rtl' ? `${formPrefixCls}-rtl` : null
  const formClassName = classnames(
    rtlCls,
    {
      [`${formPrefixCls}`]: true,
      [`${formPrefixCls}-${layout}`]: true,
    },
    className,
  )

  const formContext = useContext(FormContext)
  const [formInstance] = useForm(form)

  // 注册
  useEffect(() => {
    formContext.registerForm(name, formInstance)
    return () => {
      formContext.unregisterForm(name)
    }
  }, [name, formInstance, formContext])

  // 暴露 formInstance
  React.useImperativeHandle(ref, () => formInstance)

  // 设置默认值及回调函数
  const { setDefaultValues, setCallbacks } = formInstance.getInternalHooks(INTERNAL_HOOK_KEY)
  setDefaultValues(defaultValues)
  setCallbacks({
    onFinish,
    onFinishFailed,
    onValuesChange,
  })

  const formContextValue = React.useMemo(
    () => ({
      ...formInstance,
      labelAlign,
      labelWidth,
      wrapperWidth,
      local: formLang,
      disabled,
      vertical: layout === 'vertical',
    }),
    [formInstance, labelAlign, labelWidth, wrapperWidth, layout, formLang],
  )

  const handleSubmit = (evt: React.FormEvent<HTMLFormElement>) => {
    evt.preventDefault()
    evt.stopPropagation()

    formInstance.submit()
  }

  const handleReset = (evt: React.FormEvent<HTMLFormElement>) => {
    evt.preventDefault()
    evt.stopPropagation()

    formInstance.resetFields()
  }

  return (
    <form className={formClassName} {...restProps} onSubmit={handleSubmit} onReset={handleReset}>
      <FieldContext.Provider value={formContextValue}>{children}</FieldContext.Provider>
    </form>
  )
}

export default Form
