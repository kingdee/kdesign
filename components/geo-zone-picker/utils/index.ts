import { RegionArray } from '../interface'

export const generateRegionMap = (letterArray: string[][], regionArray: RegionArray) => {
  const regionMap: Record<string, RegionArray> = {}

  letterArray.forEach((group) => {
    const key = group.join('') // 使用 join 将数组转换为字符串，作为键
    const matchingItems = regionArray
      .filter((item) => group.includes(item.group!.toUpperCase()))
      .sort((a, b) => a.group!.localeCompare(b.group!)) // 按照 A-Z 的顺序排序

    if (matchingItems.length > 0) {
      regionMap[key] = matchingItems
    }
  })

  return regionMap
}
