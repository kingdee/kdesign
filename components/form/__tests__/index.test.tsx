import React from 'react'
import { mount, ReactWrapper } from 'enzyme'
import { sleep } from '../../../tests/utils'
import Form from '../index'
import Input from '../../input'
import Radio from '../../radio'
import Checkbox from '../../checkbox'
import Switch from '../../switch'
import Upload from '../../upload'
import Transfer from '../../transfer'
import { RangePicker } from '../../date-picker'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'

const { Item } = Form

describe('Form', () => {
  const errorSpy = jest.spyOn(console, 'error').mockImplementation()
  const warnSpy = jest.spyOn(console, 'warn').mockImplementation()
  let wrapper: ReactWrapper
  beforeEach(() => {
    wrapper = mount(
      <Form>
        <Item label="name" name="username">
          <Input />
        </Item>
      </Form>,
    )
  })

  afterEach(() => {
    wrapper = mount(
      <Form>
        <Item label="name" name="username">
          <Input />
        </Item>
      </Form>,
    )
    warnSpy.mockReset()
    errorSpy.mockReset()
  })

  afterAll(() => {
    errorSpy.mockRestore()
    warnSpy.mockRestore()
  })

  // 1. mount test
  mountTest(Form)
  mountTest(Item)

  // 2. render test
  it('renders correctly', () => {
    const LABEL_ALIGN_TYPE: any = ['left', 'right']
    const LAYOUT_TYPE: any = ['horizontal', 'vertical', 'inline']
    const wrapper = mount(
      <Form>
        <Item label="name" name={''}>
          <Input />
        </Item>
      </Form>,
    )

    expect(wrapper).toMatchSnapshot()

    LABEL_ALIGN_TYPE.forEach((labelAlign: any) => {
      wrapper.setProps({ labelAlign, labelWidth: 100 })
      wrapper.update()
      expect(wrapper).toMatchSnapshot()
    })

    LAYOUT_TYPE.forEach((layout: any) => {
      wrapper.setProps({ layout, wrappedWidth: 200 })
      wrapper.update()
      expect(wrapper).toMatchSnapshot()
    })

    expect(
      mount(
        <Form disabled>
          <Item label="name" name="" required>
            <Input />
          </Item>
        </Form>,
      ),
    ).toMatchSnapshot()
  })

  // 3. warns in component
  it('warns in component', () => {
    wrapper = mount(
      <Form>
        <Item name="name">
          <Input />
          <Input />
        </Item>
      </Form>,
    )
    expect(warnSpy).toBeCalled()
    const errorList: string[] = []
    warnSpy.mock.calls.forEach((d: string[]) => {
      errorList.push(...d)
    })

    expect(errorList.includes('Warning: [kdesign]-Form.Item: Form.Item must have only child')).toBeTruthy()
  })

  // 4. render null or undefined without errors
  describe('4. render null or undefined without errors', () => {
    it('render null or undefined without errors', () => {
      const wrapper = (
        <Form>
          {null}
          {undefined}
        </Form>
      )
      expect(wrapper).toMatchSnapshot()
    })
    it('render item null or undefined without errors', () => {
      const wrapper = (
        <Form>
          <Item name="name">
            {null}
            {undefined}
          </Item>
        </Form>
      )
      expect(wrapper).toMatchSnapshot()
    })
  })

  // 5. displayName
  it('should have displayName static property', () => {
    expect(Form.displayName).toBe('Form')
    expect(Item.displayName).toBe('Item')
  })

  // 6. class state
  describe('6. class state', () => {
    it('className', () => {
      wrapper = mount(
        <Form className={'my-form'}>
          <Item className={'my-item'} label="name" name="username">
            <Input />
          </Item>
        </Form>,
      )
      expect(wrapper.find('.my-form')).toBeTruthy()
      expect(wrapper.find('.my-item')).toBeTruthy()
    })
  })

  // 7. form event
  it('event', async () => {
    let errorInfo = null
    let result = null
    const onChange = jest.fn()
    const onValuesChange = jest.fn()
    const onFinishFailed = jest.fn((v) => {
      errorInfo = v
    })
    const onFinish = jest.fn((v) => {
      result = v
    })
    // const [form] = Form.useForm()

    const wrapper = mount(
      <Form onValuesChange={onValuesChange} onFinishFailed={onFinishFailed} onFinish={onFinish}>
        <Form.Item label="用户名" name="username" required>
          <Input onChange={onChange} />
        </Form.Item>
      </Form>,
    )

    // submit
    wrapper.find('form').first().simulate('submit')
    await sleep(100)
    expect(errorInfo).toEqual({
      values: {},
      fields: { username: [{ message: '请输入您的用户名', field: 'username' }] },
    })

    // change & submit
    wrapper.find('input').simulate('change', { target: { value: 'tom' } })
    expect(onChange).toHaveBeenCalled()
    expect(onValuesChange).toHaveBeenCalled()
    expect(onValuesChange.mock.calls[0][0]).toEqual({ username: 'tom' })
    expect(onValuesChange.mock.calls[0][1]).toEqual({ username: 'tom' })
    wrapper.find('form').first().simulate('submit')
    await sleep(100)
    expect(result).toEqual({
      values: {
        username: 'tom',
      },
    })

    // reset
    wrapper.find('form').first().simulate('reset')
    await sleep(100)
    expect(wrapper.find('input').props().value).toBe('')
  })

  // 8. config provider
  it('should config use config provider', () => {
    const config = {
      compDefaultProps: {
        Form: {
          layout: 'vertical',
          labelAlign: 'right',
          size: 'small',
        },
      },
    }
    const wrapper = mount(
      <ConfigProvider value={config}>
        <Form>
          <Item label="name" name="username">
            <Input />
          </Item>
        </Form>
      </ConfigProvider>,
    )
    expect(wrapper.find('form')).toHaveClassName('.kd-form-vertical')
    expect(wrapper.find(`.kd-form-field-label`)).toHaveClassName(`.kd-form-field-label-right`)
    expect(wrapper.find('.kd-input-size-small')).toBeTruthy()
  })

  // 9. ref test
  it('ref & FormInstance', async () => {
    const ref: any = React.createRef()
    const onFinishFailed = jest.fn()
    const onFinish = jest.fn()
    wrapper = mount(
      <Form ref={ref} defaultValues={{ firstname: 'tom' }} onFinishFailed={onFinishFailed} onFinish={onFinish}>
        <Item label="firstname" name="firstname">
          <Input />
        </Item>
        <Item label="lastname" name="lastname" required>
          <Input />
        </Item>
      </Form>,
    )

    const instance = ref.current
    // getFieldValue
    const firstname = instance.getFieldValue('firstname')
    expect(firstname).toEqual('tom')

    // getFieldsValue
    expect(instance.getFieldsValue(['firstname', 'lastname'])).toEqual({ firstname: 'tom', lastname: undefined })

    // setFieldValue
    instance.setFieldValue('firstname', 'tom1')
    expect(instance.getFieldValue('firstname')).toEqual('tom1')

    // setFieldsValue
    instance.setFieldsValue({ lastname: 'tom2' })
    expect(instance.getFieldValue('lastname')).toEqual('tom2')

    // validateFields
    const data = await instance.validateFields()
    expect(data).toEqual({ values: { firstname: 'tom1', lastname: 'tom2' } })

    // resetFields
    instance.resetFields()
    expect(instance.getFieldsValue(['firstname', 'lastname'])).toEqual({ firstname: 'tom', lastname: '' })

    // submit
    instance.submit()
    await sleep(100)
    expect(onFinishFailed).toHaveBeenCalled()
    await expect(instance.validateFields()).rejects.toEqual({
      errorInfos: {
        fields: {
          lastname: [
            {
              field: 'lastname',
              message: '请输入您的lastname',
            },
          ],
        },
        values: { firstname: 'tom', lastname: '' },
      },
      errors: [{ field: 'lastname', message: '请输入您的lastname' }],
    })

    // getFieldError
    expect(instance.getFieldError('lastname')).toEqual('请输入您的lastname')
  })

  // 10. api test
  describe('api test', () => {
    it('defaultValue', function () {
      wrapper = mount(
        <Form defaultValues={{ username: 'tom' }}>
          <Item label="name" name="username">
            <Input />
          </Item>
        </Form>,
      )
      expect(wrapper.find('input').at(0).props().value).toBe('tom')
      wrapper = mount(
        <Form>
          <Item defaultValue={'jack'} label="name" name="username">
            <Input />
          </Item>
        </Form>,
      )
      expect(wrapper.find('input').at(0).props().value).toBe('jack')
    })

    it('disabled', async () => {
      wrapper = mount(
        <Form disabled>
          <Item label="name" name="username">
            <Input />
          </Item>
        </Form>,
      )

      expect(wrapper.find('.kd-input-disabled')).toHaveLength(1)
      expect(wrapper.find('input').props().disabled).toBeTruthy()

      wrapper = mount(
        <Form>
          <Item disabled label="name" name="username">
            <Input />
          </Item>
        </Form>,
      )
      expect(wrapper.find('.kd-input-disabled')).toHaveLength(1)
      expect(wrapper.find('input').props().disabled).toBeTruthy()
    })

    it('htmlFor', () => {
      const wrapper = mount(
        <Form>
          <Item htmlFor="name" label="name" name="username">
            <Input />
          </Item>
        </Form>,
      )

      expect(wrapper.find('label').props().htmlFor).toBe('name')
    })

    it('error message & rules & validateTrigger', async () => {
      wrapper = mount(
        <Form>
          <Form.Item
            name="test"
            label="test"
            validateTrigger={'onChange'}
            rules={[
              { required: true, message: '请输入' },
              { min: 6, message: '请输入至少6位字符' },
            ]}
          >
            <Input />
          </Form.Item>
        </Form>,
      )

      wrapper.find('input').simulate('change', { target: { value: '' } })
      await sleep(200)
      wrapper.update()
      expect(wrapper.find('.kd-form-field-wrapper-message').length).toBeTruthy()
      expect(wrapper.find('.kd-form-field-wrapper-message').at(0).text()).toEqual('请输入')

      wrapper.find('input').simulate('change', { target: { value: 'king' } })
      await sleep(200)
      wrapper.update()
      expect(wrapper.find('input').at(0).props().value).toEqual('king')
      expect(wrapper.find('.kd-form-field-wrapper-message').length).toBeTruthy()
      expect(wrapper.find('.kd-form-field-wrapper-message').at(0).text()).toEqual('请输入至少6位字符')
    })
  })

  // 11. special case
  describe('special case', () => {
    it('get value_', async () => {
      let result = null
      const onFinish = jest.fn((v) => {
        result = v
      })
      const mockData = []
      for (let i = 1; i < 10; i++) {
        mockData.push({
          key: i.toString(),
          title: `选项${i}`,
          description: `选项描述${i}`,
          disabled: i % 6 === 0,
        })
      }

      const wrapper = mount(
        <Form onFinish={onFinish}>
          <Item name="radio">
            <Radio value={'radio'} checked={true} />
          </Item>
          <Item name="radiogroup">
            <Radio.Group name="radiogroup" value={'AAAA'}>
              <Radio value={'AAAA'}>A</Radio>
              <Radio value={'BBBB'}>B</Radio>
            </Radio.Group>
          </Item>
          <Item name="checkbox">
            <Checkbox value={'checkbox'} checked={true} />
          </Item>
          <Item name="checkboxgroup">
            <Checkbox.Group name="checkboxgroup" value={['AAAA']}>
              <Checkbox value={'AAAA'}>A</Checkbox>
              <Checkbox value={'BBBB'}>B</Checkbox>
            </Checkbox.Group>
          </Item>
          <Item name="switch">
            <Switch checked={true} />
          </Item>
          <Item name="upload">
            <Upload fileList={[{ uid: '123', name: '123.png', status: 'done', size: 123, type: '' } as any]} />
          </Item>
          <Item name="transfer">
            <Transfer dataSource={mockData} targetKeys={['1', '2', '3'] as any} render={(item) => item.title} />
          </Item>
          <Item name="rangepicker">
            <RangePicker value={[new Date('2000-10-01'), new Date('2000-10-31')]} />
          </Item>
        </Form>,
      )

      wrapper.find('form').first().simulate('submit')
      await sleep(100)
      expect(result).toEqual({
        values: {
          checkbox: true,
          checkboxgroup: ['AAAA'],
          radio: true,
          radiogroup: 'AAAA',
          switch: true,
          transfer: ['1', '2', '3'],
          rangepicker: [new Date('2000-10-01'), new Date('2000-10-31')],
          upload: [
            {
              name: '123.png',
              size: 123,
              status: 'done',
              type: '',
              uid: '123',
            },
          ],
        },
      })
    })

    it('instance', async () => {
      const mockInstance = {
        submit: jest.fn(),
        getInternalHooks: jest.fn().mockReturnValue({
          setDefaultValues: jest.fn(),
          setCallbacks: jest.fn(),
        }),
      }
      jest.spyOn(Form, 'useForm').mockReturnValue([mockInstance])
      const [instance] = Form.useForm()
      wrapper = mount(<Form form={instance}></Form>)
      wrapper.find('form').first().simulate('submit')
      expect(instance.submit).toHaveBeenCalled()
    })
  })
})
