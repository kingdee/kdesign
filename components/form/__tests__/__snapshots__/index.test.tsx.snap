// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders correctly 1`] = `
<Form>
  <form
    className="kd-form kd-form-vertical"
    onReset={[Function]}
    onSubmit={[Function]}
    size="middle"
  >
    <Item
      label="name"
      name=""
    >
      <div
        className="kd-form kd-form-field kd-form-field-vertical"
      >
        <FieldLabel
          value="name"
        >
          <div
            className="kd-form-field-label"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <label
              className=""
              title="name"
            >
              name
            </label>
          </div>
        </FieldLabel>
        <FieldWrapper>
          <div
            className="kd-form-field-wrapper"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <Input
              key="0"
              onChange={[Function]}
            >
              <ClearableInput
                borderType="underline"
                element={
                  <input
                    className="kd-input kd-input-size-middle kd-input-underline"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyUp={[Function]}
                    type="text"
                    value=""
                  />
                }
                focused={false}
                handleReset={[Function]}
                inputCount={null}
                inputType="input"
                onChange={[Function]}
                prefixCls="kd-input"
                size="middle"
                type="text"
                value=""
              >
                <input
                  className="kd-input kd-input-size-middle kd-input-underline"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyUp={[Function]}
                  type="text"
                  value=""
                />
              </ClearableInput>
            </Input>
          </div>
        </FieldWrapper>
      </div>
    </Item>
  </form>
</Form>
`;

exports[`Form renders correctly 2`] = `
<Form
  labelAlign="left"
  labelWidth={100}
>
  <form
    className="kd-form kd-form-vertical"
    onReset={[Function]}
    onSubmit={[Function]}
    size="middle"
  >
    <Item
      label="name"
      name=""
    >
      <div
        className="kd-form kd-form-field kd-form-field-vertical"
      >
        <FieldLabel
          value="name"
        >
          <div
            className="kd-form-field-label"
            style={
              Object {
                "width": 100,
              }
            }
          >
            <label
              className=""
              title="name"
            >
              name
            </label>
          </div>
        </FieldLabel>
        <FieldWrapper>
          <div
            className="kd-form-field-wrapper"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <Input
              key="0"
              onChange={[Function]}
            >
              <ClearableInput
                borderType="underline"
                element={
                  <input
                    className="kd-input kd-input-size-middle kd-input-underline"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyUp={[Function]}
                    type="text"
                    value=""
                  />
                }
                focused={false}
                handleReset={[Function]}
                inputCount={null}
                inputType="input"
                onChange={[Function]}
                prefixCls="kd-input"
                size="middle"
                type="text"
                value=""
              >
                <input
                  className="kd-input kd-input-size-middle kd-input-underline"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyUp={[Function]}
                  type="text"
                  value=""
                />
              </ClearableInput>
            </Input>
          </div>
        </FieldWrapper>
      </div>
    </Item>
  </form>
</Form>
`;

exports[`Form renders correctly 3`] = `
<Form
  labelAlign="right"
  labelWidth={100}
>
  <form
    className="kd-form kd-form-vertical"
    onReset={[Function]}
    onSubmit={[Function]}
    size="middle"
  >
    <Item
      label="name"
      name=""
    >
      <div
        className="kd-form kd-form-field kd-form-field-vertical"
      >
        <FieldLabel
          value="name"
        >
          <div
            className="kd-form-field-label kd-form-field-label-right"
            style={
              Object {
                "width": 100,
              }
            }
          >
            <label
              className=""
              title="name"
            >
              name
            </label>
          </div>
        </FieldLabel>
        <FieldWrapper>
          <div
            className="kd-form-field-wrapper"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <Input
              key="0"
              onChange={[Function]}
            >
              <ClearableInput
                borderType="underline"
                element={
                  <input
                    className="kd-input kd-input-size-middle kd-input-underline"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyUp={[Function]}
                    type="text"
                    value=""
                  />
                }
                focused={false}
                handleReset={[Function]}
                inputCount={null}
                inputType="input"
                onChange={[Function]}
                prefixCls="kd-input"
                size="middle"
                type="text"
                value=""
              >
                <input
                  className="kd-input kd-input-size-middle kd-input-underline"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyUp={[Function]}
                  type="text"
                  value=""
                />
              </ClearableInput>
            </Input>
          </div>
        </FieldWrapper>
      </div>
    </Item>
  </form>
</Form>
`;

exports[`Form renders correctly 4`] = `
<Form
  labelAlign="right"
  labelWidth={100}
  layout="horizontal"
  wrappedWidth={200}
>
  <form
    className="kd-form kd-form-horizontal"
    onReset={[Function]}
    onSubmit={[Function]}
    size="middle"
    wrappedWidth={200}
  >
    <Item
      label="name"
      name=""
    >
      <div
        className="kd-form kd-form-field"
      >
        <FieldLabel
          value="name"
        >
          <div
            className="kd-form-field-label kd-form-field-label-right"
            style={
              Object {
                "width": 100,
              }
            }
          >
            <label
              className=""
              title="name"
            >
              name
            </label>
          </div>
        </FieldLabel>
        <FieldWrapper>
          <div
            className="kd-form-field-wrapper"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <Input
              key="0"
              onChange={[Function]}
            >
              <ClearableInput
                borderType="underline"
                element={
                  <input
                    className="kd-input kd-input-size-middle kd-input-underline"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyUp={[Function]}
                    type="text"
                    value=""
                  />
                }
                focused={false}
                handleReset={[Function]}
                inputCount={null}
                inputType="input"
                onChange={[Function]}
                prefixCls="kd-input"
                size="middle"
                type="text"
                value=""
              >
                <input
                  className="kd-input kd-input-size-middle kd-input-underline"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyUp={[Function]}
                  type="text"
                  value=""
                />
              </ClearableInput>
            </Input>
          </div>
        </FieldWrapper>
      </div>
    </Item>
  </form>
</Form>
`;

exports[`Form renders correctly 5`] = `
<Form
  labelAlign="right"
  labelWidth={100}
  layout="vertical"
  wrappedWidth={200}
>
  <form
    className="kd-form kd-form-vertical"
    onReset={[Function]}
    onSubmit={[Function]}
    size="middle"
    wrappedWidth={200}
  >
    <Item
      label="name"
      name=""
    >
      <div
        className="kd-form kd-form-field kd-form-field-vertical"
      >
        <FieldLabel
          value="name"
        >
          <div
            className="kd-form-field-label kd-form-field-label-right"
            style={
              Object {
                "width": 100,
              }
            }
          >
            <label
              className=""
              title="name"
            >
              name
            </label>
          </div>
        </FieldLabel>
        <FieldWrapper>
          <div
            className="kd-form-field-wrapper"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <Input
              key="0"
              onChange={[Function]}
            >
              <ClearableInput
                borderType="underline"
                element={
                  <input
                    className="kd-input kd-input-size-middle kd-input-underline"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyUp={[Function]}
                    type="text"
                    value=""
                  />
                }
                focused={false}
                handleReset={[Function]}
                inputCount={null}
                inputType="input"
                onChange={[Function]}
                prefixCls="kd-input"
                size="middle"
                type="text"
                value=""
              >
                <input
                  className="kd-input kd-input-size-middle kd-input-underline"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyUp={[Function]}
                  type="text"
                  value=""
                />
              </ClearableInput>
            </Input>
          </div>
        </FieldWrapper>
      </div>
    </Item>
  </form>
</Form>
`;

exports[`Form renders correctly 6`] = `
<Form
  labelAlign="right"
  labelWidth={100}
  layout="inline"
  wrappedWidth={200}
>
  <form
    className="kd-form kd-form-inline"
    onReset={[Function]}
    onSubmit={[Function]}
    size="middle"
    wrappedWidth={200}
  >
    <Item
      label="name"
      name=""
    >
      <div
        className="kd-form kd-form-field"
      >
        <FieldLabel
          value="name"
        >
          <div
            className="kd-form-field-label kd-form-field-label-right"
            style={
              Object {
                "width": 100,
              }
            }
          >
            <label
              className=""
              title="name"
            >
              name
            </label>
          </div>
        </FieldLabel>
        <FieldWrapper>
          <div
            className="kd-form-field-wrapper"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <Input
              key="0"
              onChange={[Function]}
            >
              <ClearableInput
                borderType="underline"
                element={
                  <input
                    className="kd-input kd-input-size-middle kd-input-underline"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyUp={[Function]}
                    type="text"
                    value=""
                  />
                }
                focused={false}
                handleReset={[Function]}
                inputCount={null}
                inputType="input"
                onChange={[Function]}
                prefixCls="kd-input"
                size="middle"
                type="text"
                value=""
              >
                <input
                  className="kd-input kd-input-size-middle kd-input-underline"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyUp={[Function]}
                  type="text"
                  value=""
                />
              </ClearableInput>
            </Input>
          </div>
        </FieldWrapper>
      </div>
    </Item>
  </form>
</Form>
`;

exports[`Form renders correctly 7`] = `
<Form
  disabled={true}
>
  <form
    className="kd-form kd-form-vertical"
    onReset={[Function]}
    onSubmit={[Function]}
    size="middle"
  >
    <Item
      label="name"
      name=""
      required={true}
    >
      <div
        className="kd-form kd-form-field kd-form-field-vertical"
      >
        <FieldLabel
          requiredMark={true}
          value="name"
        >
          <div
            className="kd-form-field-label"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <label
              className="kd-form-field-label-required-mark"
              title="name"
            >
              name
            </label>
          </div>
        </FieldLabel>
        <FieldWrapper>
          <div
            className="kd-form-field-wrapper"
            style={
              Object {
                "width": undefined,
              }
            }
          >
            <Input
              disabled={true}
              key="0"
              onChange={[Function]}
            >
              <ClearableInput
                borderType="underline"
                disabled={true}
                element={
                  <input
                    className="kd-input kd-input-size-middle kd-input-underline kd-input-disabled"
                    disabled={true}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyUp={[Function]}
                    type="text"
                    value=""
                  />
                }
                focused={false}
                handleReset={[Function]}
                inputCount={null}
                inputType="input"
                onChange={[Function]}
                prefixCls="kd-input"
                size="middle"
                type="text"
                value=""
              >
                <input
                  className="kd-input kd-input-size-middle kd-input-underline kd-input-disabled"
                  disabled={true}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onFocus={[Function]}
                  onKeyUp={[Function]}
                  type="text"
                  value=""
                />
              </ClearableInput>
            </Input>
          </div>
        </FieldWrapper>
      </div>
    </Item>
  </form>
</Form>
`;
